fix audio and video

Kitrinx opened the pull request:
Sorg: If you have time, please double check to make sure i'm not doing anything wrong in video changes. You have a lot more experience with this than me.

-Fix aspect ratio for NES. Although NES and SNES only draw 256 pixels, they pad the image to 280 pixels so that all of the visible pixels are inside the overscan. Therefor we scale to 4:3 from the correct resolution of 280x240. I adjusted the overscan calculation to match. I confirmed that now the image on a CRT and on HDMI matches exactly what you would see from a real NES.

-Fix frame timer in APU. It was causing a variety of notes to end early, audible in Ducktales, and probably many other things.

-Fix square/mix channel balance. The old LUT wasn't well balanced, this new one passes the APU_MIXER tests much better and matches the original NES audio levels.

- Fix crackling sounds caused by supersonic triangle cancellation.

-Re-balanced mapper audio mixing around fixed APU balance.

-Fix MMC5 audio bug I introduced in last commit.

sorgelig commented:
I didn't check it yet, but when 4:3 AR is activated on HDMI, you don't obliged to use exactly 4 and 3 for ARX/ARY. You may choose any numbers which will compensate video to get correct 4:3 AR on HDMI.

Kitrinx commented:
I am aware that the AR can be changed, but it seemed more sound for the scaling to start with the actual frame dimensions rather than starting with altered frame dimensions before applying scaling.

sorgelig commented:
but 280x240 is not 4:3 from pixels count if we assume the pixels are square.

Kitrinx commented:
correct, 280x240 is not 4:3, the 280x240 image was scaled to 4:3 by the TV, since the ppu was actually outputting that resolution. So if we do the same, take a 280x240 image and scale it to 4:3, the aspect ratio is correct as it would be scaled by a CRT television. The linearity test in 280p test suite passes this perfectly. Alternatively you could take 256x240 and try to figure out what the correct aspect ratio is for it to make it show up correctly, but I find it easier to follow what the original system did.

sorgelig commented:
Ok, i see now.
I'm using 4:3 monitor and linearity test looks perfect to me. I see it cuts less in "hide overscan" mode. Some games like Boulder Dash use wide edge areas to construct the field, so it's more visible now, but that's ok. I have no preference here.

I think what is really missing in NES is PAL mode. I know you are from NTSC region, but may be you could try to implement PAL mode?

Kitrinx commented:
when I refactored the top level file I changed the way the clock dividers work  to play better with PAL, but it will be tricky. There's differences in the APU and PPU as well, and the RAM timing has to play well with two independent clock dividers. I am working my way up to it.

sorgelig commented:
Tell me more about required clocks. I can add PLL reconfiguration.

Kitrinx commented:
This has all the info about pal vs NTSC:
https://wiki.nesdev.com/w/index.php/Cycle_reference_chart

The master clock for PAL mode should be 26.6017125mhz
